### ISSUE TITLE
Golang client library - New signature for ContainerWait makes it difficult to use

### ISSUE THREAD

**Issue by nkashy1** [moby/moby#40430]
**Description**

The first issue is somewhat superficial. [`ContainerWait`](https://github.com/moby/moby/blob/master/client/container_wait.go) used to return `(int64, error)` but now returns `(<-chan container.ContainerWaitOKBody, <-chan error)`. This places a lot more burden on the caller than the previous signature did - to select from the channels.

The second issue is more concerning - `ContainerWait` seems to have changed its behavior from the versions in which it had the previous signature. It used to be the case that calling `ContainerWait` on a container which was not running anymore would return immediately. Now, when `ContainerWait` runs on an exited container (with `container.WaitConditionNotRunning` as its `condition` argument), it blocks indefinitely.

I don't think this is a bug, since the comments in [`client/container_wait.go`](https://github.com/moby/moby/blob/master/client/container_wait.go) suggest that this is the expected behavior.

It does make it difficult to use `ContainerWait` when the `ContainerStart` call was made in a separate scope.

The [tests](https://github.com/moby/moby/blob/master/integration/container/wait_test.go) don't provide any help on the expected usage of `ContainerWait` in this setting.

My immediate workaround was to listen on the channels in a goroutine which I kill pending a `ContainerInspect` if the container is already in a not running state. I plan to replace this with a simple poll to make the code more readable. Others may have use cases where polling is not a reasonable alternative, though.

Would it be reasonable to provide a utility function which provides the functionality of the old `ContainerWait`?

*PS:* I apologize for not providing library versions - they are a little difficult to reason about. (Related to: https://github.com/moby/moby/issues/39302, which also mentions `ContainerWait` in passing.)

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.1
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        74b1e89e8a
 Built:             Thu Jul 25 21:21:35 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.1
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       74b1e89e8a
  Built:            Thu Jul 25 21:20:09 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 19
  Running: 0
  Paused: 0
  Stopped: 19
 Images: 438
 Server Version: 19.03.1
 Storage Driver: aufs
  Root Dir: /var/lib/docker/aufs
  Backing Filesystem: extfs
  Dirs: 475
  Dirperm1 Supported: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.4.0-157-generic
 Operating System: Ubuntu 16.04.2 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 7.51GiB
 Name: donkey
 ID: JFDR:HRGK:723X:BTLM:YTAI:R3WW:4T5E:CC6Y:UEKX:JND2:WV6I:2WYY
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: simiotics
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
WARNING: the aufs storage-driver is deprecated, and will be removed in a future release.
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Laptop running Ubuntu 16.04.2

**Comment by thaJeztah** [comment]
Yes, that API because rather involved; changes were made for `docker run --rm` to have the "auto remove after exit" handled by the daemon (before that change, the container would be removed by the CLI, causing various issues); later changes were needed to keep compatibility with `docker swarm` (standalone or "classic" swarm); see https://github.com/moby/moby/issues/32242 and https://github.com/moby/moby/pull/32237

**Comment by nkashy1** [comment]
Fair enough, it is much more important to properly integrate the CLI with the different API implementations in this case.

Perhaps we can accommodate all use cases by adding a new value for `container.WaitCondition`? Something like `"already-exited-or-not-running"? If the container with the given name is not in a running state, the `container.ContainerWaitOKBody` channel will immediately return the value with the appropriate exit code.

By the way, it isn't clear to me what the difference is between `container.WaitConditionNotRunning` and `container.WaitConditionNextExit`. The tests don't seem to exercise these conditions. I would be happy to add some tests if someone would explain the conditions to me.
